Publish configuration

huntc opened the pull request:
This commit permits bundle configuration to be published to bintray, as per the existing ability to publish bundles.

In addition a licensing check has been relaxed as that checks belongs to the server side.

huntc committed:
Publish configuration

This commit permits bundle configuration to be published to bintray, as per the existing ability to publish bundles.

In addition a licensing check has been relaxed as that checks belongs to the server side.

fsat commented:
@huntc - I have some questions:
1. Can I confirm if we run `sbt configuration:publish` - it will publish only the configuration, not the bundle?
2. For project like Elasticsearch where you have the following bundle and configuration:

`conductr-elasticsearch-v1-8d209c71effc12f925a07a3e0f8ca15ba1ac8a6c0cbaf0d80cb5c70593213c2d.zip`
`elasticsearch-prod-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87.zip`

The bundle will sit under `typesafehub/bundles/conductr-elasticsearch` in Bintray.

Will the config sit under `typesafehub/bundles/conductr-elasticsearch`?

Also, based my understanding on reading the code, the artefact version will be `conductr-elasticsearch-elasticsearch-prod-v1-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87` - is this correct?

fsat commented:
@huntc - I did a quick test trying to publish ES config.

```
Felixs-MacBook-Pro-2:conductr felixsatyaputra$ sbt conductrElasticsearch/configuration:publish
[info] Loading project definition from /Users/felixsatyaputra/workspace/typesafe-fsat/conductr/project/project
[info] Loading project definition from /Users/felixsatyaputra/workspace/typesafe-fsat/conductr/project
[info] Set current project to root (in build file:/Users/felixsatyaputra/workspace/typesafe-fsat/conductr/)
[info] Bundle configuration has been created: /Users/felixsatyaputra/workspace/typesafe-fsat/conductr/conductr-elasticsearch/target/bundle-configuration/elasticsearch-prod-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87.zip
[info] Checking package for felix-satyaputra/conductr-elasticsearch-elasticsearch-prod@v1-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87
java.lang.RuntimeException: was not able to find or create a package for felix-satyaputra in Repo(felix-satyaputra,bundle) named conductr-elasticsearch-elasticsearch-prod
    at scala.sys.package$.error(package.scala:27)
    at bintray.BintrayRepo.ensurePackage(BintrayRepo.scala:49)
    at com.typesafe.sbt.bintraybundle.BintrayBundle$$anonfun$settings$5.apply(BintrayBundle.scala:66)
    at com.typesafe.sbt.bintraybundle.BintrayBundle$$anonfun$settings$5.apply(BintrayBundle.scala:36)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:237)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
[error] (conductrElasticsearch/configuration:publish) was not able to find or create a package for felix-satyaputra in Repo(felix-satyaputra,bundle) named conductr-elasticsearch-elasticsearch-prod
[error] Total time: 2 s, completed 11/11/2016 3:11:41 PM
```

Unfortunately it seems publishing not to the default org `typesafehub` and repo `bundle`.

Also, I've noticed the package name is `conductr-elasticsearch-elasticsearch-prod`. So the user will be resolving this config using `conductr-elasticsearch-elasticsearch-prod`, which I don't think will be as nice.

```
conduct load conductr-elasticsearch conductr-elasticsearch-elasticsearch-prod 
```

vs

```
conduct load conductr-elasticsearch elasticsearch-prod
```

I'm thinking perhaps it makes sense for the bundle configuration to be published to repo `bundle-configuration`, and the configuration name is kept as is.

All of these would mean change to the CLI though.

fsat reviewed (changes requested):
Bundle configuration needs to be published to `typesafehub` by default, and having a default repo specified.

We can discuss whether `bundle` or `bundle-configuration` should be used as the default repo.

huntc commented:
> Can I confirm if we run sbt configuration:publish - it will publish only the configuration, not the bundle?

Confirmed.

> For project like Elasticsearch where you have the following bundle and configuration:
> conductr-elasticsearch-v1-8d209c71effc12f925a07a3e0f8ca15ba1ac8a6c0cbaf0d80cb5c70593213c2d.zip
> elasticsearch-prod-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87.zip
> 
> The bundle will sit under typesafehub/bundles/conductr-elasticsearch in Bintray.
> 
> Will the config sit under typesafehub/bundles/conductr-elasticsearch?

The config will sit wherever you declare it to sit. I've updated the README to state that you must set the repo, org etc. explicitly for config (no values are inherited). This is because config is precious.

> Also, based my understanding on reading the code, the artefact version will be conductr-elasticsearch-elasticsearch-prod-v1-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87 - is this correct?

I've now changed this so that the config in the above will become just:

```
elasticsearch-prod-v1-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87
```

fsat commented:
Publishing config works now.

```
Felixs-MacBook-Pro-2:conductr felixsatyaputra$ sbt conductrElasticsearch/configuration:publish
[info] Loading project definition from /Users/felixsatyaputra/workspace/typesafe-fsat/conductr/project/project
[info] Loading project definition from /Users/felixsatyaputra/workspace/typesafe-fsat/conductr/project
[info] Set current project to root (in build file:/Users/felixsatyaputra/workspace/typesafe-fsat/conductr/)
[info] Bundle configuration has been created: /Users/felixsatyaputra/workspace/typesafe-fsat/conductr/conductr-elasticsearch/target/bundle-configuration/elasticsearch-prod-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87.zip
[info] Checking package for typesafe/elasticsearch-prod@v1-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87
[info] Staging typesafe/elasticsearch-prod@v1-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87
[info] /Users/felixsatyaputra/workspace/typesafe-fsat/conductr/conductr-elasticsearch/target/bundle-configuration/elasticsearch-prod-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87.zip was uploaded to typesafe/elasticsearch-prod@v1-6437dbf56fe76a60e22d014637a32938d98286cebdae22abb66c2a93e7eaea87
[success] Total time: 4 s, completed 11/11/2016 4:20:32 PM
```

Config is published fine in bintray.

<img width="928" alt="screen shot 2016-11-11 at 4 22 28 pm" src="https://cloud.githubusercontent.com/assets/1665611/20205231/404d002a-a82b-11e6-9481-2382d1993b6c.png">

fsat commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Permit bundle configurations to be deployed in the same manner as bundles.](https://trello.com/c/GQePhnYn/402-permit-bundle-configurations-to-be-deployed-in-the-same-manner-as-bundles)
